modin/backends/pandas/query_compiler.py

```diff
@@ -509,6 +509,83 @@ def is_monotonic_decreasing(self):
+        Return a QueryCompiler of Series containing counts of unique values.
+
+        Returns
+        -------
+        PandasQueryCompiler
+        """
+        if kwargs.get("bins", None) is not None:
+            new_modin_frame = self._modin_frame._apply_full_axis(
+                0, lambda df: df.squeeze(axis=1).value_counts(**kwargs)
+            )
+            return self.__constructor__(new_modin_frame)
+
+        def map_func(df, *args, **kwargs):
+            return df.squeeze(axis=1).value_counts(**kwargs)
+
+        def reduce_func(df, *args, **kwargs):
+            normalize = kwargs.get("normalize", False)
+            sort = kwargs.get("sort", True)
+            ascending = kwargs.get("ascending", False)
+            dropna = kwargs.get("dropna", True)
+
+            try:
+                result = df.squeeze(axis=1).groupby(df.index, sort=False).sum()
+            except (ValueError):
+                result = df.copy().squeeze(axis=1).groupby(df.index, sort=False).sum()
+
+            if not dropna and np.nan in df.index:
+                result = result.append(
+                    pandas.Series(
+                        [df.squeeze(axis=1).loc[[np.nan]].sum()], index=[np.nan]
+                    )
+                )
+            if normalize:
+                result = result / df.squeeze(axis=1).sum()
+
+            result = result.sort_values(ascending=ascending) if sort else result
+
+            def sort_index_for_identical_values(result, ascending):
+                is_range = False
+                is_end = False
+                i = 0
+                new_index = np.array([], dtype=type(result.index))
+                while i < len(result):
+                    j = i
+                    if i < len(result) - 1:
+                        while result[result.index[i]] == result[result.index[i + 1]]:
+                            i += 1
+                            if is_range is False:
+                                is_range = True
+                            if i == len(result) - 1:
+                                is_end = True
+                                break
+                    if is_range:
+                        new_index = np.concatenate(
+                            (
+                                new_index,
+                                sorted(result.index[j : i + 1], reverse=not ascending),
+                            )
+                        )
+                        if is_end:
+                            break
+                        is_range = False
+                    else:
+                        new_index = np.concatenate(
+                            (new_index, np.array([result.index[j]]))
+                        )
+                    i += 1
+                return pandas.DataFrame(result, index=new_index)
+
+            return sort_index_for_identical_values(result, ascending)
+
+        return MapReduceFunction.register(map_func, reduce_func, preserve_index=False)(
```

Comment:
Thinking on how to structure this to keep the query compiler clean of these low-level implementation details, what if we created a file `backends/pandas/kernels.py`, then added these functions there and import? I have been thinking about how to keep things clean, maybe it is better left to another PR, but what do you think on this?

Reply:
Yes, this is great thought. I think, a file `backend/pandas/kernels.py` that will contain these implementation details is what we need to keep query compiler clean. Importing function from that file into query compiler is a very nice solution.